src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs

```diff
@@ -198,26 +198,36 @@ public static unsafe string TryGetNameInfo(IPAddress addr, out SocketError socke
         {
             byte* buffer = stackalloc byte[Interop.Sys.NI_MAXHOST + 1 /*for null*/];
 
-            // TODO #2891: Remove the copying step to improve performance. This requires a change in the contracts.
-            byte[] addressBuffer = addr.GetAddressBytes();
-
-            int error;
-            fixed (byte* rawAddress = &addressBuffer[0])
-            {
-                error = Interop.Sys.GetNameInfo(
-                    rawAddress,
-                    unchecked((uint)addressBuffer.Length),
-                    addr.AddressFamily == AddressFamily.InterNetworkV6 ? (byte)1 : (byte)0,
-                    buffer,
-                    Interop.Sys.NI_MAXHOST,
-                    null,
-                    0,
-                    Interop.Sys.GetNameInfoFlags.NI_NAMEREQD);
+            byte isIPv6;
+            int rawAddressLength;
+            if (addr.AddressFamily == AddressFamily.InterNetwork)
+            {
+                isIPv6 = 0;
+                rawAddressLength = IPAddressParserStatics.IPv4AddressBytes;
+            }
+            else
+            {
+                isIPv6 = 1;
+                rawAddressLength = IPAddressParserStatics.IPv6AddressBytes;
             }
 
+            byte* rawAddress = stackalloc byte[rawAddressLength];
```

Comment:
Nothing here was required to change, but I saw it as I was auditing stackallocs and decided to address the TODO, switching from allocating a byte[] to using a stackalloc and the new span-based IPAddress support.